File: BE/CodeGenX64/isel_tester.cc
// (c) Robert Muth - see LICENSE for more info

#include "Base/serialize.h"
#include "BE/CodeGenX64/isel_gen.h"
#include "BE/CodeGenX64/regs.h"
#include "BE/CpuX64/symbolic.h"
#include "Util/parse.h"

#include <iostream>

namespace {

using namespace cwerg;
using namespace cwerg::base;

const char* OpTypeString(Handle op) {
  switch (op.kind()) {
    default:
      return "_";
    case RefKind::CONST:
      return EnumToString(ConstKind(Const(op)));
    case RefKind::REG:
      return EnumToString(RegKind(Reg(op)));
  }
}

void HandleIns(Ins ins, const code_gen_x64::EmitContext& ctx) {
  std::cout << "INS: ";
  InsRenderToAsm(ins, &std::cout);
  std::cout << "  [";
  const char* sep = "";
  for (unsigned i = 0; i < InsOpcode(ins).num_operands; ++i) {
    std::cout << sep << OpTypeString(InsOperand(ins, i));
    sep = " ";
  }
  std::cout << "]\n";

  const code_gen_x64::Pattern* pat = code_gen_x64::FindMatchingPattern(ins);
  ASSERT(pat != nullptr, "");
  std::cout << "PAT: reg:[";
  sep = "";
  for (unsigned i = 0; i < InsOpcode(ins).num_operands; ++i) {
    std::cout << sep << EnumToString(pat->type_curbs[i]);
    sep = " ";
  }
  std::cout << "]  op:[";
  sep = "";
  for (unsigned i = 0; i < InsOpcode(ins).num_operands; ++i) {
    std::cout << sep << code_gen_x64::EnumToString(pat->op_curbs[i]);
    sep = " ";
  }
  std::cout << "]\n";

  std::vector<std::string> ops;
  for (unsigned i = 0; i < pat->length; ++i) {
    const code_gen_x64::InsTmpl& tmpl = pat->start[i];
    x64::Ins cpu_ins = code_gen_x64::MakeInsFromTmpl(tmpl, ins, ctx);
    ops.clear();
    std::string_view name = x64::InsSymbolize(cpu_ins, true, false, &ops);
    std::cout << "    " << name;
    for (const std::string& op : ops) std::cout << " " << op;
    std::cout << "\n";
  }
}

void Process(std::istream* input) {
  code_gen_x64::InitCodeGenX64();
  std::vector<char> data = SlurpDataFromStream(input);

  Unit unit = UnitParseFromAsm("unit", {data.data(), data.size()},
                               code_gen_x64::GetAllRegs());
  code_gen_x64::EmitContext ctx;
  // UnitRenderToAsm(unit, &std::cout);
  for (Fun fun : UnitFunIter(unit)) {
    FunFinalizeStackSlots(fun);
    ctx.scratch_cpu_reg =  code_gen_x64::GPR_REGS[0];
    for (Bbl bbl : FunBblIter(fun)) {
      for (Ins ins : BblInsIter(bbl)) {
        std::cout << "\n";
        HandleIns(ins, ctx);
      }
    }
  }
}

}  // namespace

int main(int argc, const char* argv[]) {
  InitStripes(1);
  Process(&std::cin);
  return 0;
}

File: BE/CodeGenX64/isel_tester.py
#!/bin/env python3
"""Testing helper for table driven code selection"""

from BE.CpuX64 import symbolic
from BE.Base import serialize
from BE.Base import ir
from BE.CodeGenX64 import isel_tab
from BE.CodeGenX64 import regs

from typing import Any
import sys


def OpToStr(op: Any) -> str:
    return str(op)


def OpTypeStr(op: Any) -> str:
    if isinstance(op, ir.Reg):
        return op.kind.name
    elif isinstance(op, ir.Const):
        return op.kind.name
    else:
        return "_"


def HandleIns(ins: ir.Ins, ctx: regs.EmitContext):
    print("INS: " + serialize.InsRenderToAsm(
        ins).strip() + f"  [{' '.join(OpTypeStr(o) for o in ins.operands)}]")

    pattern = isel_tab.FindMatchingPattern(ins)
    assert pattern
    print(
        f"PAT: reg:[{' '.join(a.name for a in pattern.type_constraints)}]  "
        f"op:[{' '.join(a.name for a in pattern.op_curbs)}]")
    for tmpl in pattern.emit:
        x64ins = tmpl.MakeInsFromTmpl(ins, ctx)
        name, ops = symbolic.InsSymbolize(x64ins, True)
        print(f"    {name} {' '.join(ops)}")


def Translate(fin):
    unit = serialize.UnitParseFromAsm(fin, cpu_regs=regs.CPU_REGS_MAP)
    for fun in unit.funs:
        ctx = regs.EmitContext(0, 0, 0)
        ctx.scratch_cpu_reg = regs.CPU_REGS_MAP["rax"]
        fun.FinalizeStackSlots()
        for bbl in fun.bbls:
            for ins in bbl.inss:
                print()
                HandleIns(ins, ctx)


if __name__ == "__main__":
    Translate(sys.stdin)
